model-optimizer/extensions/front/tf/while_ext.py

```diff
@@ -0,0 +1,169 @@
+            assert (body_graph.has_node(src_id))
+            edge_attrs = {
+                'out': src_port,
+                'in': dst_port,
+                'name': src_id,
+                'fw_tensor_debug_info': [(src_id, src_id)],
+                'in_attrs': ['in', 'name'],
+                'out_attrs': ['out', 'name'],
+                'data_attrs': ['fw_tensor_debug_info']
+            }
+            body_graph.add_edge(src_id, id, **edge_attrs)
+
+    # create Result nodes in the body graph
+    for output in body_graph_proto['output_arg']:
+        src_id = body_graph_proto['ret'][output.name].split(":")[0]
+        if src_id not in body_parameter_names:
+            src_id = prefix_name + src_id
+        src_port = 0 if len(body_graph_proto['ret'][output.name].split(":")) == 1\
+            else int(body_graph_proto['ret'][output.name].split(":")[-1])
+        assert body_graph.has_node(src_id), 'The body graph does not contain output with name "{}"'.format(
+            src_id)
+        body_results.append(Node(body_graph, add_opoutput(body_graph, src_id, src_port, False)))
+
+
+class WhileExtractor(FrontExtractorOp):
+    op = 'StatelessWhile'
+    enabled = True
+
+    @classmethod
+    def extract(cls, loop_node):
+        Loop.update_node_stat(loop_node, {})
+        loop_name = loop_node.soft_get('name', loop_node.id)
+
+        # check that required body and condition functions exist in the graph library
+        main_graph = loop_node.graph
+        body_graph_name = loop_node.pb.attr['body'].func.name
+        cond_graph_name = loop_node.pb.attr['cond'].func.name
+        assert 'library' in main_graph.graph, 'The graph does not contain a library that is required ' \
+                                                     'by node with name "{}".'.format(loop_name)
+        library_graph = main_graph.graph['library']
+
+        assert body_graph_name in library_graph, 'The library does not contain a function with name "{}" ' \
+                                                 'that is required by node ' \
+                                                 'with name "{}".'.format(body_graph_name, loop_name)
+        body_graph_proto = library_graph[body_graph_name]
+
+        assert cond_graph_name in library_graph, 'The library does not contain a function with name "{}" ' \
+                                                 'that is required by node ' \
+                                                 'with name "{}".'.format(cond_graph_name, loop_name)
+        cond_graph_proto = library_graph[cond_graph_name]
+
+        body_graph = Graph()
+        # fill the body graph except library since the library is enough to have in the main graph
+        for attr_key in main_graph.graph.keys():
+            if attr_key != 'library':
+                body_graph.graph[attr_key] = copy.deepcopy(main_graph.graph[attr_key])
+        loop_node['body'] = body_graph
```

Comment:
What will happen if we have a While inside While? Where the library of inner While is stored: in a main graph or in the body of the outer While?

Reply:
Agree to add a library into a body graph because it will be recursively extracted in this case

Reply:
I've created separate extractors for `While` and `StatelessWhile` and described the need.